source/common/router/scoped_rds.cc

```diff
@@ -100,72 +107,220 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
+    // In the case if factory_context_.initManager() is uninitialized, RDS is already paused either
+    // by Server init or LDS init.
+    factory_context_.clusterManager().adsMux().pause(
+        Envoy::Config::TypeUrl::get().RouteConfiguration);
+    resume_rds = std::make_unique<Cleanup>([this, &noop_init_manager, version_info] {
+      // For new RDS subscriptions created after listener warming up, we don't wait for them to warm
+      // up.
+      Init::WatcherImpl noop_watcher(
+          // Note: we just throw it away.
+          fmt::format("SRDS ConfigUpdate watcher {}:{}", name_, version_info),
+          []() { /*Do nothing.*/ });
+      noop_init_manager->initialize(noop_watcher);
+      // New RDS subscriptions should have been created, now lift the floodgate.
+      // Note in the case of partial acceptance, accepted RDS subscriptions should be started
+      // despite of any error.
+      factory_context_.clusterManager().adsMux().resume(
+          Envoy::Config::TypeUrl::get().RouteConfiguration);
+    });
   }
+  std::vector<std::string> exception_msgs;
+  bool any_applied = addOrUpdateScopes(
+      added_resources,
+      (noop_init_manager == nullptr ? factory_context_.initManager() : *noop_init_manager),
+      version_info, exception_msgs);
+  any_applied = removeScopes(removed_resources, version_info) || any_applied;
+  ConfigSubscriptionCommonBase::onConfigUpdate();
+  if (any_applied) {
+    setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
+  }
+  stats_.config_reload_.inc();
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(fmt::format("Error adding/updating scoped route(s): {}",
+                                     StringUtil::join(exception_msgs, ", ")));
+  }
+}
 
-  std::unordered_set<std::string> resource_names;
-  for (const auto& scoped_route : scoped_routes) {
-    if (!resource_names.insert(scoped_route.name()).second) {
+void ScopedRdsConfigSubscription::onRdsConfigUpdate(const std::string& scope_name,
+                                                    RdsRouteConfigSubscription& rds_subscription) {
+  auto iter = scoped_route_map_.find(scope_name);
+  ASSERT(iter != scoped_route_map_.end(),
+         fmt::format("trying to update route config for non-existing scope {}", scope_name));
+  auto new_scoped_route_info = std::make_shared<ScopedRouteInfo>(
+      envoy::api::v2::ScopedRouteConfiguration(iter->second->configProto()),
+      std::make_shared<ConfigImpl>(rds_subscription.routeConfigUpdate()->routeConfiguration(),
+                                   factory_context_, false));
+  applyConfigUpdate([new_scoped_route_info](ConfigProvider::ConfigConstSharedPtr config)
+                        -> ConfigProvider::ConfigConstSharedPtr {
+    auto* thread_local_scoped_config =
+        const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
+    thread_local_scoped_config->addOrUpdateRoutingScope(new_scoped_route_info);
```

Comment:
This lambdai is the updateCb, which means it is executed on each worker thread.
Is there any race condition on concurrently modify the referenced object of `thread_local_scoped_config` ?

Reply:
ThreadLocalConfig contains a thread local `shared_ptr`. The shared_ptr has a thread local copy but I am not sure if ScopedConfigImpl `object` is duplicated on each thread. 

Thank you for the explanation and the confirm

Reply:
I see. 
so SRDS subscription is a DeltaConfigSubscriptionInstance, it holds one ThreadLocal Config instance in each thread, which points to a per-thread ScopedConfigImpl.

The ConfigUpdateCb definition in config-provider-framework is defined:
std::function<ConfigProvider::ConfigConstSharedPtr(ConfigProvider::ConfigConstSharedPtr)>; 
which implies we want to do some immutable programming here, but here we know what we are doing, so we cast off the constness. -_-